### PR TITLE
[Add Example] Create reproducibility_recipes.py

### PR DESCRIPTION
**Pull request by Frightera**
This quick Keras recipe demonstrates how Keras initializers behave, and shows what's needed to achieve identical results in subsequential runs in model training process.

#1332

### CODE REVIEW

File: examples/keras_recipes/reproducibility_recipes.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+from tensorflow import keras
+from tensorflow.keras import layers
+
+# Set the seed using keras.utils.set_random_seed. This will set:
+# 1) `numpy` seed
+# 2) `tensorflow` random seed
+# 3) `python` random seed
+keras.utils.set_random_seed(812)
+
+# This will make TensorFlow ops as deterministic as possible.
```
Comment (LukeWood) [inline examples/keras_recipes/reproducibility_recipes.py:34]: perhaps mention the performance loss that follows from this to explain why its not on by default

File: examples/keras_recipes/reproducibility_recipes.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+initializing the weights of layer variables. The following built-in initializers
+are available as part of `tf.keras.initializers`:
+"""
+to_be_removed = ["Initializer", "Ones", "Zeros", "Identity"]
+initializers_list = [
+    string
+    for string in dir(keras.initializers)
+    if string[0].isupper() and string not in to_be_removed
+]
+print(initializers_list)
```
Comment (LukeWood) [inline examples/keras_recipes/reproducibility_recipes.py:51]: can you explain this code a bit more in comments?  I'm not following the logic of "why" some of these are picked and others are not!

The goal is to remove these from the initializers module?

File: examples/keras_recipes/reproducibility_recipes.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        if "seed" in initializer_signature.parameters:
+            result = initializer(seed=42)(shape=(3, 3))
+        else:
+            result = initializer()(shape=(3, 3))
+        results[initializer_name].append(result)
+
+# Check if the results are equal.
+all_equal = True
+for initializer_name, initializer_results in results.items():
+    if not tf.experimental.numpy.allclose(
```
Comment (LukeWood) [inline examples/keras_recipes/reproducibility_recipes.py:79]: so we're not quite removing modules, we're just checking to ensure that these output identical results?  I think that is reasonable!

**Review by LukeWood — COMMENTED** [review]
Perhaps a useful guide - I think we could greatly simplify this code and make it very concise by removing a lot of the module inspection code and some of the more complex logic loops!

**Comment by Frightera** [comment]
@LukeWood Thanks for the review, I simplified those parts, can you take a look at when you have time?

File: examples/keras_recipes/reproducibility_recipes.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+for initializer in initializers_list:
+    print(f"Running {initializer}")
+
+    for iteration in range(2):
+        # In order to get same results across multiple runs from an initializer,
+        # you need to specify a seed value. Note that this is not related to
+        # keras.utils.set_random_seed or tf.config.experimental.enable_op_determinism.
+        # If you comment those lines, you will still get the same results.
+        result = initializer(seed=42)(shape=(1, 1)).numpy()
```
Comment (fchollet) [inline examples/keras_recipes/reproducibility_recipes.py:75]: You could cast to float instead of using `numpy()` since it's a single element

File: examples/keras_recipes/reproducibility_recipes.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+    # Resize the image
+    image = tf.image.resize(image, (32, 32))
+
+    return image, label
+
+
+# Prepare the datasets, batch-map --> vectorized operations
+train_data = (
+    train_ds.shuffle(buffer_size=len(train_images))
```
Comment (fchollet) [inline examples/keras_recipes/reproducibility_recipes.py:207]: Data order during training is important so shuffle should be explained

**Review by fchollet — COMMENTED** [review]
Thank you for the PR! Looking good! 👍

**Comment by Frightera** [comment]
Hi @fchollet, thank you for your comments. Now I provided more explanation for the `shuffle` process.

**Review by fchollet — APPROVED** [review]
Thank you! Everything looks good to me. You can add the generated files.

**Comment by Frightera** [comment]
@fchollet Done! I think this is ready to be merged 😄 Here's a snippet:

image

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the great contribution! 👍
